composeApp/src/androidMain/kotlin/io/github/couchtracker/db/user/ExternalUserDb.kt

```diff
@@ -0,0 +1,171 @@
+import io.github.couchtracker.db.common.DbPath
+import io.github.couchtracker.db.lastModifiedInstant
+import io.github.couchtracker.db.toDocumentFile
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import kotlinx.datetime.Instant
+import kotlin.coroutines.CoroutineContext
+
+/**
+ * A [UserDb] that is managed in an external file.
+ *
+ * The [transaction] function of this class takes care of copying the external file to internal storage before opening it, and then copying
+ * it back, when necessary.
+ */
+class ExternalUserDb private constructor(
+    private val user: User,
+    private val cachedDb: DbPath,
+    private val externalDb: DocumentFile,
+    private val appData: AppData,
+) : UserDb() {
+
+    override suspend fun <T> doTransaction(context: Context, block: DatabaseTransaction<TransactionResult<T>>): UserDbResult<T> {
+        // Check if we have an up-to-date cached copy of the database. If not, copy it to internal storage
+        val externalLastModified = externalDb.lastModifiedInstant()
+        if (!isCachedDatabaseUpToDate(externalLastModified)) {
+            Log.i(LOG_TAG, "External database is not up to date or doesn't exist, copying to internal storage")
+            externalDb.uri.copyToInternal(context, cachedDb.file)?.let { return it }
+            updateCachedLastModified(externalLastModified)
+        }
+
+        // Execute transaction on cached file
+        val dbResult = openAndUseDatabase(
+            context = context,
+            dbPath = cachedDb,
+            onCorrupted = {
+                // this probably means that the external file is not a valid SQLite file
+                // we'll just delete the cached file
+                cleanCached()
+            },
+            block = block,
+        )
+
+        if (dbResult is UserDbResult.Completed.Success) {
+            // The transaction completed successfully on the local file
+
+            if (dbResult.result.edited) {
+                // Check the last modified date of the external file again, as it might have changed while we were running the transaction
+                val currentExternalLastModified = externalDb.lastModifiedInstant()
+                if (currentExternalLastModified != externalLastModified) {
+                    Log.i(
+                        LOG_TAG,
+                        "External file changed while transaction was running! " +
+                            "Old last modified: $externalLastModified, current: $currentExternalLastModified",
+                    )
+                    // We cannot copy the newly modified file over, or we'll end overwriting external changes
+                    // In this case we'll need to run the transaction again
+                    return doTransaction(context, block)
+                }
+
+                // Copy the now edited cached file back to its rightful place
+                cachedDb.file.copyToExternal(context, externalDb.uri)?.let { return it }
```

Comment:
These early returns leave the internal db in a modified state. So, from the user perspective, they performed an action, they got an error, but the change got applied in the app. 
Then, with the next action they take (assuming now the external storage part works), the app will override the internal file with the external one, effectively undoing the previous operation.

I don't have any fix for this bug, as we can't use db transactions, since the changes have to be committed by the time we get here. Any ideas?